src/main/java/htsjdk/variant/vcf/VCFIteratorBuilder.java

```diff
@@ -0,0 +1,210 @@
+ * 
+ * Example:
+ * 
+ * <pre>
+ * VCFIterator r = new VCFIteratorBuilder().open(System.in);
+ * while (r.hasNext()) {
+ *     System.out.println(r.next());
+ * }
+ * r.close();
+ * </pre>
+ * 
+ * @author Pierre Lindenbaum / @yokofakun
+ * @see {@link VCFIterator}
+ *
+ */
+
+public class VCFIteratorBuilder {
+
+    /**
+     * creates a VCF iterator from an input stream It detects if the stream is a
+     * BCF stream or a GZipped stream.
+     * 
+     * @param in
+     *            inputstream
+     * @return the VCFIterator
+     * @throws RuntimeIOException
+     */
+    @SuppressWarnings("resource")
+    public VCFIterator open(InputStream in) throws IOException {
+        if (in == null) {
+            throw new IllegalArgumentException("input stream is null");
+            }
+        // try to read the BCF header
+        final BufferedInputStream bufferedinput = new BufferedInputStream(in, 100);
+        bufferedinput.mark(100);
+        final BCFVersion bcfVersion = BCFVersion.readBCFVersion(bufferedinput);
+        bufferedinput.reset();
+
+        //this is VCF or VCF.gz
+        if (bcfVersion == null) {
+            // try to read GZIP header
+            // see http://stackoverflow.com/questions/4818468/how-to-check-if-inputstream-is-gzipped
+            final byte[] signature = new byte[2];
+            bufferedinput.mark(signature.length);
+            final int len = bufferedinput.read(signature); // read the signature
+            bufferedinput.reset(); // push back the signature to the stream
+            if (len == signature.length && signature[0] == (byte) 0x1f && signature[1] == (byte) 0x8b) { 
+                //it's gzip, decode the gzip input
+                in = new GZIPInputStream(bufferedinput);
+            } else {
+                // it's plain VCF
+                in = bufferedinput;
+            }
+            return new VCFReaderIterator(in);
+        } else {
+            return new BCFInputStreamIterator(bufferedinput);
+        }
+    }
+
+    /**
+     * creates a VCF iterator from an URI It detects if the stream is a BCF
+     * stream or a GZipped stream.
+     * 
+     * @param uri
+     *            the URI
+     * @return the VCFIterator
+     * @throws OException
+     */
+    public VCFIterator open(final String uri) throws IOException {
+        if (IOUtil.isUrl(uri)) {
+            try {
+                final URL url = new URL(uri); // this will throw if its not a
+                                              // url
+                return open(url.openStream());
+            } catch (final MalformedURLException e) {
+                // ignore
+            }
+        }
+        return open(new File(uri));
+    }
+
+    /**
+     * creates a VCF iterator from a File
+     * 
+     * @param file
+     *            the file (can be bcf, vcf, vcf.gz)
+     * @return the VCFIterator
+     * @throws IOException
+     */
+    public VCFIterator open(final File file) throws IOException {
```

Comment:
For the `java.nio.Path` method, if the previous change is made, the simpler option will be:

```java
public VCFIterator open(final Path path) throws IOException {
    return open(new SeekablePathStream(path, null));
}
```